packages/multichain/src/middlewares/multichainMethodCallValidator.ts

```diff
@@ -0,0 +1,114 @@
+import { MultiChainOpenRPCDocument } from '@metamask/api-specs';
+import type { JsonRpcMiddleware } from '@metamask/json-rpc-engine';
+import { rpcErrors } from '@metamask/rpc-errors';
+import { isObject } from '@metamask/utils';
+import type { Json, JsonRpcError, JsonRpcParams } from '@metamask/utils';
+import type {
+  ContentDescriptorObject,
+  MethodObject,
+  OpenrpcDocument,
+} from '@open-rpc/meta-schema';
+import dereferenceDocument from '@open-rpc/schema-utils-js/build/dereference-document';
+import { makeCustomResolver } from '@open-rpc/schema-utils-js/build/parse-open-rpc-document';
+import type { Schema, ValidationError } from 'jsonschema';
+import { Validator } from 'jsonschema';
+
+const transformError = (
+  error: ValidationError,
+  param: ContentDescriptorObject,
+  got: unknown,
+) => {
+  // if there is a path, add it to the message
+  const message = `${param.name}${
+    error.path.length > 0 ? `.${error.path.join('.')}` : ''
+  } ${error.message}`;
+
+  return {
+    code: -32602, // TODO: could be a different error code or not wrapped in json-rpc error, since this will also be wrapped in a -32602 invalid params error
+    message,
+    data: {
+      param: param.name,
+      path: error.path,
+      schema: error.schema,
+      got,
+    },
+  };
+};
+
+const v = new Validator();
+
+const dereffedPromise = dereferenceDocument(
+  MultiChainOpenRPCDocument as unknown as OpenrpcDocument,
+  makeCustomResolver({}),
+);
+const multichainMethodCallValidator = async (
+  method: string,
+  params: JsonRpcParams | undefined,
+) => {
+  const dereffed = await dereffedPromise;
+
+  const methodToCheck = dereffed.methods.find(
+    (m) => (m as unknown as ContentDescriptorObject).name === method,
+  );
+
+  const errors: JsonRpcError[] = [];
+  if (
+    !methodToCheck ||
+    !isObject(methodToCheck) ||
+    !('params' in methodToCheck)
+  ) {
+    return [
+      {
+        code: -32601,
+        message: 'The method does not exist / is not available.',
+        data: {
+          method,
+        },
+      },
+    ];
+  }
+  // check each param and aggregate errors
+  (methodToCheck as unknown as MethodObject).params.forEach((param) => {
+    const p = param as ContentDescriptorObject;
+    if (!isObject(params)) {
+      errors.push({
+        code: -32602,
+        message: 'Invalid method parameter(s).',
+      });
+      return;
+    }
+    const paramToCheck = params[p.name];
+
+    const result = v.validate(paramToCheck, p.schema as unknown as Schema, {
+      required: p.required,
+    });
+    if (result.errors) {
+      errors.push(
+        ...result.errors.map((e) => {
+          return transformError(e, p, paramToCheck) as JsonRpcError;
+        }),
+      );
+    }
+  });
+  if (errors.length > 0) {
+    return errors;
+  }
+  // feels like this should return true to indicate that its valid but i'd rather check the falsy value since errors
```

Comment:
Perhaps this is indicating that this function should return two things rather a single thing? `{ isValid: true } | { isValid: false; errors: JsonRpcError[] }`

Reply:
made this just return an array always and I do an array length check downstream

https://github.com/MetaMask/core/pull/4813/commits/e77e87f08d18a2ab618bba2133a4ba00f3b81190